arangod/Replication2/ReplicatedLog/Supervision.cpp

```diff
@@ -499,36 +530,43 @@ auto checkReplicatedLog(LogTarget const& target,
     return WriteEmptyTermAction(minTerm);
   }
 
+  // barrier
+  if (!current.leader || !current.leader->committedParticipantsConfig ||
+      current.leader->committedParticipantsConfig->generation !=
+          plan.participantsConfig.generation) {
+    return EmptyAction(
+        "Waiting for participants configuration to be committed by the "
+        "leader.");
+  }
+
   // Check whether a participant was added in Target that is not in Plan.
   // If so, add it to Plan.
   if (auto participant = getAddedParticipant(
           target.participants, plan.participantsConfig.participants)) {
     return AddParticipantToPlanAction(participant->first, participant->second);
   }
 
-  // If a participant is in Plan but not in Target, gracefully
-  // remove it
-  if (auto maybeParticipant = getRemovedParticipant(
-          target.participants, plan.participantsConfig.participants)) {
-    auto const& [participantId, planFlags] = *maybeParticipant;
-
-    // We do not ever remove a leader
-    if (participantId != leader.serverId) {
-      // If the participant is not allowed in Quorum it is safe to remove it
-      if (not planFlags.allowedInQuorum and
-          current.leader->committedParticipantsConfig->generation ==
-              plan.participantsConfig.generation) {
-        return RemoveParticipantFromPlanAction(participantId);
-      } else if (planFlags.allowedInQuorum) {
-        // A participant can only be removed without risk,
-        // if it is not member of any quorum
-        auto newFlags = planFlags;
-        newFlags.allowedInQuorum = false;
-        return UpdateParticipantFlagsAction(participantId, newFlags);
-      } else {
-        // still waiting
-        return EmptyAction("Waiting for participants config to be committed");
-      }
+  // TODO this is a hack
+  //      doParticipantRemoval will determine a planned participant that is not
+  //      in target and not the leader.
+  //      We can only execute this removal if we do not violate writeConcern
+  //      otherwise we need to be able to continue checking flags, because
+  //      the only way participants will become eligible for removal is
+  //      by flags being updated
+  auto action = doParticipantRemoval(target.participants,
+                                     plan.participantsConfig.participants,
+                                     leader.serverId);
+
+  auto removePending = false;
+  if (action.has_value()) {
+    auto canExecuteRemove =
+        target.participants.size() <=
```

Comment:
Why the =? Updating the flags to allowed in quorum false makes this participant NOT ok, and you end up with too few not ok participants.